pkg/boot/linux/load_linux_arm64.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build (linux && arm64) || (linux && riscv64)
+// +build linux,arm64 linux,riscv64
+
+package linux
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"syscall"
+
+	"github.com/u-root/u-root/pkg/boot/image"
+	"github.com/u-root/u-root/pkg/dt"
+	"github.com/u-root/u-root/pkg/uio"
+	"golang.org/x/sys/unix"
+)
+
+// loadFDT loads a flateen device tree blob.
+//
+// It first tries to load it from given file path, then
+// from /sys/firmware/fdt. If still failed, it resorts
+// to read /proc/device-tree, and flatten it.
+func loadFDT(dtb string) (*dt.FDT, error) {
+	// TODO(10000TB): impl this.
+}
+
+// sanitizeFDT cleanups chosen node from given FDT.
+func sanitizeFDT(*dt.FDT) (*fdt.Node, error) {
+	// Clear old entries in case we've already been through kexec to get
+	// to this instance of runtime.
+	chosen, _ := fdt.NodeByName("chosen")
+	if chosen == nil {
+		return nil, fmt.Errorf("no /chosen node in device tree")
+	}
+	for _, property := range []string{"linux,elfcorehdr", "linux,usable-memory-range", "kaslr-seed", "rng-seed", "linux,initrd-start", "linux,initrd-end"} {
+		chosen.RemoveProperty(property)
+	}
+
+	return chosen, nil
+}
+
+func mmap(f *os.File) (data []byte, ummap func() error, err error) {
+	fd := f.Fd()
+	s, err := f.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("stat error: %w", err)
+	}
+	if s.Size() == 0 {
+		return nil, fmt.Errorf("cannot mmap zero-len file")
+	}
+	d, err := unix.Mmap(int(fd), 0, int(s.Size()), syscall.PROT_READ, syscall.MAP_PRIVATE)
+	if err != nil {
+		return nil, fmt.Errorf("mmap failed: %w", err)
+	}
+
+	ummap = func() error {
+		return unix.Munmap(d)
+	}
+
+	return d, ummap, nil
+}
+
+// KexecLoad loads arm64 Image, with the given ramfs and kernel cmdline.
+func KexecLoad(kernel, ramfs *os.File, cmdline string, opts KexecOptions) error {
+	var err error
+	// kmem is a struct holding kexec segments.
+	//
+	// It has routines to work with physical memory
+	// ranges.
+	var kmem *Memory
+	var kernelRange, ramfsRange, dtbRange, trampolineRange Range
+
+	fdt, err := loadFDT(opts.DTB)
+	if err != nil {
+		return fmt.Errorf("loadFDT(%s) = %v", opts.DTB, err)
+	}
+
+	chosen, err := sanitizeFDT(fdt)
+	if err != nil {
+		return fmt.Errorf("sanitizeFDT(%v) = %v", fdt, err)
+	}
+
+	// Prepare segments.
+	kmem = &Memory{}
+	Debug("Try parsing memory map...")
+	if err := kmem.ParseMemoryMapFromFDT(fdt); err != nil {
+		return fmt.Errorf("ParseMemoryMapFromFDT(%v): %v", fdt, err)
+	}
+
+	var kernelBuf []byte
+	if opts.MmapKernel {
+		Debug("Mmapping kernel to virtual buffer...")
+		kernelBuf, ummap, err := mmap(kernel)
+		if err != nil {
+			return fmt.Errorf("mmap kernel: %v", err)
+		}
+		defer func() {
```

Comment:
I don't understand the point of the unmap. The kernel will do this anyway when kexec exits. Is there some reason for this?

Reply:
You are right, if it kexec-s through.

I was thinking what if not, and any harm it would do if not doing unmapping. But I don't have a good such example. Those mapping probably just cost a few KBs of layered page tables in current kernel at its max, anyway ? and we mmap as Read and Private, no harm if references to those page cache are not cleaned up. But they are tmpfs page cache anyway.

ah.. I don't have a good reason :-) ...merely as a habit to unmap, kind like a file open and close thing -- does not hurt

Reply:
Consider the failure cases -- trying to netboot a large image that fails (say due to not enough memory), so we fallback to diskboot. If we leave the netboot images mmapped, the memory usage does not drop if/when we clean up the tmpfs files, and it's possible we won't have enough memory to complete the diskboot (which we expect to work if not for the netboot).

Reply:
ugh...I seem to recall it is usually reboot in those OOM cases. e.g. we called into kexec sys call, and kernel try to file load, and them OOM, and reboot. Did it fall back to diskboot ?

I can imagine it falls back to diskboot, if OOM happened before making the kexec sys call. 

(Just from my re-collections, did not double check those recent OOM console logs)

Reply:
I don't see the advantage of the mmap. It seems to add complexity with no real advantage. What's the benefit?

Reply:
actually I don't see the argument around diskboot as a recovery for mmap. In the diskboot case, you read the whole file in. Linux has a unified page and buffer cache. IIUC, you won't do any better or worse with the mmap than the diskboot. There are cases where mmap can  be a big win, but I'd like to at least see some measurement to indicate this is one of them.

Reply:
For motivation of doing mmap against the kernel and initrd file, than `ioutil.ReadAll` (for example), it was found to save  memory. The understanding is that latter will allocate addition physical page frames to store kernel and initrd, and backing the virtual buffer, while former would directly map the virtual buffer against the tmpfs page cache. I guess we agree on this, iirc.

I think the only question is what Ron asked in #1, "I don't understand the point of unmap...". agree ?

>  If we leave the netboot images mmapped, the memory usage does not drop if/when we clean up the tmpfs files, and it's possible we won't have enough memory to complete the diskboot (which we expect to work if not for the netboot).

If we hit OOM during netboot, and exits, the OS image (downloaded) stays in tmpfs. There is no common logic in u-root to purge that file. I guess @dillow 's point is that when users do cleanups to delete the netboot OS image in tmpfs in their init, so that free the memory for later diskboot case (If this is a real user case, is not clear to me, but probably should assume such creative is possible). 

> actually I don't see the argument around diskboot as a recovery for mmap. In the diskboot case, you read the whole file in. Linux has a unified page and buffer cache. IIUC, you won't do any better or worse with the mmap than the diskboot. There are cases where mmap can be a big win, but I'd like to at least see some measurement to indicate this is one of them.

diskboot look for boot images from local disk, not diskboot-ing the same netboot OS downloaded from netboot. Diskboot can decides to do disk I/O read, or mmap the OS from disk, but that is different payload from netboot, and the page cache created are additional on top of the netboot payload from tmpfs.